mdal/mdal_data_model.cpp

```diff
@@ -332,3 +337,293 @@ size_t MDAL::Mesh::faceVerticesMaximumCount() const
+  {
+    values.year--;
+    values.month += 12;
+  }
+
+  int A = values.year / 100;
+  int B = A / 4;
+  int C = 2 - A + B;
+  int E = int( 365.25 * ( values.year + 4716 ) );
+  int F = int( 30.6001 * ( values.month + 1 ) );
+  double julianDay = C + values.day + E + F - 1524.5;
+
+  mJulianTime = int64_t( julianDay * 24 * 3600 * 1000 +
+                         ( values.hours ) * 3600 * 1000 + //the formula set the day with hours at 00h
+                         values.minutes * 60 * 1000 +
+                         values.seconds * 1000 );
+}
+
+void MDAL::DateTime::setWithGregorianJulianCalendarValues( MDAL::DateTime::DateTimeValues values )
+{
+  //https://quasar.as.utexas.edu/BillInfo/JulianDatesG.html
+
+  int C = 0;
+  int Y = values.year;
+  int M = values.month;
+  int D = values.day;
+  if ( M <= 2 )
+  {
+    Y--;
+    M += 12;
+  }
+  if ( values.year > 1582 ||
+       ( values.year == 1582 && ( values.month > 10 || ( values.month == 10 && values.day >= 15 ) ) ) ) //gregorian calendar
+  {
+    int A = Y / 100;
+    int B = A / 4;
+    C = 2 - A + B;
+  }
+
+  int E = int( 365.25 * ( Y + 4716 ) );
+  int F = int( 30.6001 * ( M + 1 ) );
+  double julianDay = C + D + E + F - 1524.5;
+
+  mJulianTime = int64_t( julianDay * 24 * 3600 * 1000 +
+                         ( values.hours ) * 3600 * 1000 + //the formula set the day with hours at 00h
+                         values.minutes * 60 * 1000 +
+                         values.seconds * 1000 );
+}
+
+std::string MDAL::DateTime::toString( MDAL::DateTime::DateTimeValues values ) const
+{
+  std::string yearStr = std::to_string( values.year );
+
+  int miliseconds = int( ( values.seconds - int( values.seconds ) ) * 1000 + 0.5 );
+  std::string msStr;
+  if ( miliseconds > 0 )
+  {
+    if ( miliseconds < 10 )
+      msStr = prependZero( std::to_string( miliseconds ), 3 );
+    else if ( miliseconds < 100 )
+      msStr = prependZero( std::to_string( miliseconds ), 2 );
+
+    msStr = std::string( "," ).append( msStr );
+  }
+
+  std::string strDateTime = prependZero( std::to_string( values.year ), 4 ) + "-" +
+                            prependZero( std::to_string( values.month ), 2 ) + "-" +
+                            prependZero( std::to_string( values.day ), 2 ) + "T" +
+                            prependZero( std::to_string( values.hours ), 2 ) + ":" +
+                            prependZero( std::to_string( values.minutes ), 2 ) + ":" +
+                            prependZero( std::to_string( int( values.seconds ) ), 2 ) +
+                            msStr; /// TODO use another way to translate seconds with fraction aprt
+
+  return strDateTime;
+}
+
+MDAL::Duration MDAL::DateTime::operator-( const MDAL::DateTime &other ) const
+{
+  if ( !mValid || !other.mValid )
+    return Duration();
+  return Duration( mJulianTime - other.mJulianTime );
+}
+
+MDAL::Duration::Duration(): mDuration( 0 )
+{}
+
+MDAL::Duration::Duration( double duration, MDAL::Duration::Unit unit )
```

Comment:
in general this is true, however UNIDATA (formats with CF/UGRID convention) are a bit tricky. see 

```udunits defines the periods as fixed multiples of seconds. The non-obvious ones are:

day = 86400.0 seconds
week = 7 days = 604800.0 seconds
month = year/12 = 2629743.831225 seconds
year = 3.15569259747E7 seconds
```

https://www.unidata.ucar.edu/software/netcdf-java/current/CDM/CalendarDateTime.html

you would need to have some special parameter in Duration DateTime to specify it is CF input and not normal one, so you can use different coefficient

Reply:
I am not sure to understand... 
The CF/UGRID convention is consistent with day=86400 seconds, and weeks=7 days, so why treat this unit differently.
For year/month, we can adopt the CF convention or not. Right now, the user has to choose how to convert year/month unit by converting himself before calling this constructor.

Reply:
no problem, just to keep it in mind when working with formats based on CF.

Reply:
Finally, I have implemented year and month too. This is necessary if CF driver meet this time unit. Duration class could adopts UNIDATA. Not sure... I have to think about...